### PR TITLE
[clang-tidy]detecting conversion directly by `make_unique` and `make_shared` in bugprone-optional-value-conversion

### PR DESCRIPTION
**Pull request by HerrCai0907**
Inspired by #110964

### CODE REVIEW

File: clang-tools-extra/clang-tidy/bugprone/OptionalValueConversionCheck.cpp [hunk 1]
```diff
@@ -12,20 +12,44 @@
...
       Finder, Builder);
 }
 
+AST_MATCHER_P2(Expr, constructFrom, Matcher<QualType>, TypeMatcher,
+               Matcher<Expr>, ArgumentMatcher) {
+  std::array<StringRef, 2> NameList{
+      "::std::make_unique",
+      "::std::make_shared",
+  };
+  return expr(anyOf(
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/OptionalValueConversionCheck.cpp:36]: that's not efficient, new matchers will be created constantly.
replace this with simple function that returns this matcher.

File: clang-tools-extra/clang-tidy/bugprone/OptionalValueConversionCheck.cpp [hunk 2]
```diff
@@ -12,20 +12,44 @@
...
-              InnerMatcher) {
+AST_MATCHER_P(QualType, hasCleanType, Matcher<QualType>, InnerMatcher) {
   return InnerMatcher.matches(
       Node.getNonReferenceType().getUnqualifiedType().getCanonicalType(),
       Finder, Builder);
 }
 
+AST_MATCHER_P2(Expr, constructFrom, Matcher<QualType>, TypeMatcher,
+               Matcher<Expr>, ArgumentMatcher) {
+  std::array<StringRef, 2> NameList{
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/OptionalValueConversionCheck.cpp:32]: make constexpr and move outside.

**Review by PiotrZSL — COMMENTED** [review]
Note: Probably also std::make_optional could be supported.

Please improve performance - avoid creating temporary matchers on every match.
Except above, looks fine.

**Comment by firewave** [comment]
Can the logic for implementing this also be used to address https://github.com/llvm/llvm-project/issues/86447#issuecomment-2016943524?

**Comment by HerrCai0907** [comment]
> Can the logic for implementing this also be used to address https://github.com/llvm/llvm-project/issues/86447#issuecomment-2016943524?

I think this kind model of stl can be reused in other check. but i have no idea to make it more generic.

**Comment by HerrCai0907** [comment]
> Note: Probably also std::make_optional could be supported.

I will do it in separated PR since std::make_optional is more complex stuff. see #119554

**Review by 5chmidti — APPROVED** [review]

**Pull request merged**